### PR TITLE
6889 Add relationships on single transaction

### PR DESCRIPTION
**Pull request by APriestman**

### CODE REVIEW

**Review by rcordovano — CHANGES_REQUESTED** [review]

File: bindings/java/src/org/sleuthkit/datamodel/CommunicationsManager.java [hunk 1]
```diff
@@ -381,17 +383,50 @@ public void addRelationships(AccountFileInstance sender, List<AccountFileInstanc
...
+				break;
+			default:
+				throw new TskCoreException("Unknown DB Type: " + db.getDatabaseType().name());
+		}		
+		
+		CaseDbTransaction trans = db.beginTransaction();	
+		try {
+			SleuthkitCase.CaseDbConnection connection = trans.getConnection();
+			PreparedStatement preparedStatement = connection.getPreparedStatement(query, Statement.NO_GENERATED_KEYS);
+
```
Comment (rcordovano) [inline bindings/java/src/org/sleuthkit/datamodel/CommunicationsManager.java:405]: I believe that both CaseDatabase and PreparedStatement are AutoCloseable so the Codacy issue regarding closing resources could be resolved with a try-with-resources.

**Review by rcordovano — CHANGES_REQUESTED** [review]

File: bindings/java/src/org/sleuthkit/datamodel/CommunicationsManager.java [hunk 2]
```diff
@@ -381,17 +383,50 @@ public void addRelationships(AccountFileInstance sender, List<AccountFileInstanc
...
+			case POSTGRESQL:
+				query = "INSERT " + query + " ON CONFLICT DO NOTHING";
+				break;
+			case SQLITE:
+				query = "INSERT OR IGNORE " + query;
+				break;
+			default:
+				throw new TskCoreException("Unknown DB Type: " + db.getDatabaseType().name());
+		}		
+
```
[the remark this answers is not in the document]
Comment (APriestman) [reply bindings/java/src/org/sleuthkit/datamodel/CommunicationsManager.java:400]: No - the transaction gets the single user case lock and releases it in commit/rollaback
private CaseDbTransaction(SleuthkitCase sleuthkitCase, CaseDbConnection connection) throws TskCoreException {
	this.connection = connection;
	this.sleuthkitCase = sleuthkitCase;
	try {
		this.connection.beginTransaction();
	} catch (SQLException ex) {
		throw new TskCoreException("Failed to create transaction on case database", ex);
	}
	sleuthkitCase.acquireSingleUserCaseWriteLock();
}

File: bindings/java/src/org/sleuthkit/datamodel/CommunicationsManager.java (hunk 1 above)
Comment (APriestman) [reply bindings/java/src/org/sleuthkit/datamodel/CommunicationsManager.java:405]: I can put the CaseDbConnection part in a try with resources, but the prepared statement really shouldn't be. The transaction holds onto those for future use and closes them itself when it is closed (yes that's at the same time in our case, but we still shouldn't be manually closing it).

Comment (APriestman) [reply bindings/java/src/org/sleuthkit/datamodel/CommunicationsManager.java:405]: Actually the CaseDbTransaction closes the connection itself, so it seems pointless to put that in a try-with-resources. It's part of the CaseDbTransaction object, along with its prepared statements.

**Review by rcordovano — APPROVED** [review]

**Pull request merged**